CARDS-1103: As a user, I can see answers from other forms in the form I am filling in

acrowthe opened the pull request:
Add in reference questions, which take a questionnaire and quest as a property.
If there is a form for the current subject with that questionnaire and question,
  copy of that answer once when the reference answer is created.
  
To create a new reference question, set the question's `entryMode` to reference and `dataType` to any other question dataType.

2 testing forms were created in the --test runmode, `ReferenceTestUser` and ReferenceTestCopied`. To test these forms, fill in a single copy of ReferenceTestUser for a subject, then create a ReferenceTestCopied form for that same subject

To test:
- Reference questions
- Backend computed questions (via the backend computed questions test form)
- FormAnswerCopyProcessor and SubjectAnswerCopyProcessor (test via PROMS run mode, instructions available [here](https://github.com/data-team-uhn/cards/pull/964) with merged implementation available [here](https://github.com/data-team-uhn/cards/pull/984))

acrowthe committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in

Add in refernce questions, which take a questionnaire and quest as a property.
If there is a form for the current subject with that questionnaire and question,
  copy of that answer once when the reference answer is created.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/api/QuestionnaireUtils.java:
```diff
@@ -119,6 +119,15 @@
+     * Check if the given node is a Question node for a reference question.
+     *
+     * @param node the node to check, a JCR Node, may be {@code null}
+     * @return {@code true} if the node is not {@code null}, is computed and is of type {@code cards:Question},
```
```suggestion
     * @return {@code true} if the node is not {@code null}, is of type {@code cards:Question}, and is a reference,
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ReferenceQuestion.jsx:
```diff
@@ -0,0 +1,182 @@
+// Other options are passed to the <question> widget
+//
+// Sample usage, given question_a and question_b are number questions:
+//<ReferenceQuestion
```
This needs to be updated.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ReferenceQuestion.jsx:
```diff
@@ -0,0 +1,182 @@
+  const [muiInputProps, changeMuiInputProps] = useState({});
+  const [isFormatted, changeIsFormatted] = useState(false);
+
+  const form = useFormReaderContext();
```
Not used, should be removed.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ReferenceQuestion.jsx:
```diff
@@ -0,0 +1,182 @@
+        valueType={answerType}
+        pageActive={pageActive}
+        {...rest}
+        />
```
I wonder if we can remove this, since the answer is always generated on the server.

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/179163471-57cecd9d-2c66-4594-9ee9-95898d4af10a.png)

acrowthe committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
- Fix rebase conflict causing failure to obtain service user
- Fix incorrect code comments
- Remove unused code from front end reference questions
- Remove unneeded access from service user ACL

acrowthe committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
- Fix rebase issue with static access warning preventing build

acrowthe commented:
Known issue: trying to reference a date question results in no answer for that reference question

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ReferenceQuestion.jsx:
```diff
@@ -0,0 +1,159 @@
+    text: PropTypes.string.isRequired,
+    description: PropTypes.string,
+    questionnaire: PropTypes.string.isRequired,
+    question: PropTypes.string.isRequired,
```
Unused.

```suggestion
```

sdumitriu committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
Add support for multivalued and date fields

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ReferenceQuestion.jsx:
```diff
@@ -0,0 +1,159 @@
+  let answerType, answerNodeType = `cards:${capitalizedDataType}Answer`, newFieldType;
+  switch (dataType) {
+    case "boolean":
+      answerType = "Long"; // Long, not Boolean
+      break;
+    case "date":
+      newFieldType = DateQuestionUtilities.getFieldType(dateFormat);
+      setFieldType(newFieldType);
+      switch (newFieldType) {
+        case "long":
+          answerType = "Long";
+          break;
+        case "string":
+          answerType = "Date";
+          break;
+        default:
+          answerType = "Date";
+          break;
+      }
+      break;
+    case "long":
+    case "double":
+    case "decimal":
+      answerType = capitalizedDataType;
+      break;
+    case "vocabulary":
+      answerType = "String";
+      break;
+    case "time":
+      answerType = "String";
+      newFieldType = Time.timeQuestionFieldType(dateFormat);
+      setFieldType(newFieldType);
+      break;
+    case "text":
+    default: // Fallthrough default to string reference
+      answerType = "String";
+      answerNodeType = "cards:TextAnswer"
+      break;
+  }
+
```
All this seems useless since you're not using any of these variables.

```suggestion
```

acrowthe replied:
Unfortunately, <Answer> needed to be added back to fix computed and conditionals which rely on reference questions, so this is needed again

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ReferenceAnswersEditor.java:
```diff
@@ -98,8 +102,7 @@ protected boolean isQuestionNodeMatchingType(Node node)
     {
-        if (this.isFormNode && "questionnaire".equals(after.getName()))
-        {
+        if (this.isFormNode && "questionnaire".equals(after.getName())) {
```
Code style fixes should have been in a separate commit.

sdumitriu reviewed (commented):
I fixed the copying of date answers, so the value is present in the form, but the UI fails to display it properly. The issue is that the raw value in the JSON is a full ISO date+time, but the input expects it to be in the specified format.

Copying multivalued answers also failed (I fixed this). However, the display of multivalued fields is not ideal, it is a comma separated list like `value a,value b,value c`.

Both these UI issues suggest that the UI part is doing unnecessary work trying to reproduce what all the other question types are doing. Is it possible to forward the call to the right `AnswerComponent` by pretending this is a readonly non-referenced answer? Something like:

```
let mockQuestionDefinition = clone questionDefinition and set entryMode="user", display="readonly";
let ProperQuestionDisplay = AnswerComponentManager.getAnswerComponent(mockQuestionDefinition);
return <ProperQuestionDisplay {...props}/>;
```

marta- commented:
> Both these UI issues suggest that the UI part is doing unnecessary work trying to reproduce what all the other question types are doing. Is it possible to forward the call to the right `AnswerComponent` by pretending this is a readonly non-referenced answer? Something like:
> 
> ```
> let mockQuestionDefinition = clone questionDefinition and set entryMode="user", display="readonly";
> let ProperQuestionDisplay = AnswerComponentManager.getAnswerComponent(mockQuestionDefinition);
> return <ProperQuestionDisplay {...props}/>;
> ```

I like this idea in principle, but:
- there's no `display="readonly" atm, we could enforce `isEdit={false}` instead to display the value like in view mode
- for most questions, view mode display depends on `displayedValue` being present in the JSON, which will not be the case here

marta- commented:
> * for most questions, view mode display depends on `displayedValue` being present in the JSON, which will not be the case here

We should have a label processor for reference questions that generates the `displayedValue` based on the source question definition.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -38,6 +38,15 @@
             }
+          },
+          "reference": {
+            "questionnaire": "string",
+            "question": "string",
```
I think having a single field `question` which holds the path of the reference question node will make out lives easier going forward. Among other things, it makes it easier to access the question definition from java code, and it should result in full expansion in the form JSON, again providing easy access to the question definition.

acrowthe committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
- Re-add Answer node to fix conditionals and computed questions that rely on referenceQuestions

acrowthe committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
- Remove `Questionnaire` property from reference question, instead `question` will store a full question path

sdumitriu committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
Re-enable the date questions since the values are actually working, just the display in edit mode is broken

sdumitriu committed:
[cleanup] Simplify code

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ReferenceAnswersEditor.java:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    @SuppressWarnings("checkstyle:NestedIfDepth")
+    private Object getAnswer(NodeState form, String questionPath)
```
Something similar is happening in `io.uhndata.cards.forms.internal.serialize.FormAnswerCopyProcessor.findForm(Node, Node)`, with the extra feature that it also supports related subjects, not just the same subject. We should make that code reusable by moving it in `FormUtils` and use it in both places.

sdumitriu committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
Deduplicated code by adding FormUtils methods for finding answers in other forms

acrowthe committed:
CARDS-1103: As a user, I can see answers from other forms in the form I am filling in
- Simplify ReferenceQuestion frontend to manually add answer to formContext rather than using <Answer>

sdumitriu committed:
[misc] Fixed error logged when creating a new visit information form

sdumitriu commented:
I'm not sure this works in background jobs, e.g. prems import + reference questions. Should we spend a bit of time fixing this scenario @marta- ?

marta- commented:
> I'm not sure this works in background jobs, e.g. prems import + reference questions. Should we spend a bit of time fixing this scenario @marta- ?

For prems there is one use case for reference questions: in surveys, populate the "Hospital" question (first question in each survey) with the "Location" from Visit information. I tested this on the `CARDS-1781_on_CARDS-1103` branch and it doesn't seem to work. We should fix it, but maybe not in the first iteration. The Heracles usecase ( #1143 ) appears to work fine.

marta- commented:
> > I'm not sure this works in background jobs, e.g. prems import + reference questions. Should we spend a bit of time fixing this scenario @marta- ?
> 
> For prems there is one use case for reference questions: in surveys, populate the "Hospital" question (first question in each survey) with the "Location" from Visit information. I tested this on the `CARDS-1781_on_CARDS-1103` branch and it doesn't seem to work, i.e. the Hospital doesn't get populated in survey forms generated in the backend when importing visits. When creating the survey forms manually everything works as expected. We should fix it, but maybe not in the first iteration. The Heracles usecase ( #1143 ) appears to work fine.

marta- reviewed (approved):
The code and the current functionality (based on light testing) seem fine as far as the Heracles project is concerned. Further improvements may be needed, but they are not blockers at this time.

marta- commented:
Marking as incomplete because of the following missing parts:
- [ ] label processor for reference questions that replicates the displayed value of the source
- [ ]  listener that captures the changes to the source answer and updates the copied answer
- [ ] fix whatever prevents the ref answer from being populated when surveys are generated by the VisitChangeListener in proms/prems

marta- commented:
> Marking as incomplete because of the following missing parts:
> 
> * [ ]  label processor for reference questions that replicates the displayed value of the source
> * [ ]   listener that captures the changes to the source answer and updates the copied answer
> * [ ]  fix whatever prevents the ref answer from being populated when surveys are generated by the VisitChangeListener in proms/prems

However, none of the above should block merging this PR.
